### ISSUE TITLE
same request produces different response bodies (body inconsistently returns instance of Hash or String)

### ISSUE THREAD

**Issue by ksilex** [lostisland/faraday#1388]
## Basic Info

* Faraday Version: 1.6.0
* Ruby Version: 2.7.4

## Issue description
seems like when requests done in parallel response body doesn't get parsed

## Steps to reproduce
```ruby
def run1
    req = Faraday.new(config[:url]) do |conn|
      conn.response :json, content_type: %r{application/json}
    end
    req.get('/') do |conn|
      conn.body = body
    end
end
```
![Screenshot_2](https://user-images.githubusercontent.com/66538478/151222255-ef517d5c-e8ca-4b51-9d63-012e5bbff79c.png)

also seen similar problem here: https://stackoverflow.com/questions/67182790/error-response-not-parsed-by-faraday-despite-adding-json-middleware
except no middlewares were used

**Comment by olleolleolle** [comment]
👋 Can you attempt with a higher version of 1.x series?

**Comment by olleolleolle** [comment]
**Update**: Here's a Bundler inline example, which allows you to pick versions used and have the example actually runnable.

I guessed at some information, so I may be off: this program only returned String for me.

```ruby
#!/usr/bin/env ruby

require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem "faraday", "= 1.6.0"
  gem "faraday_middleware"
end

def run1
  url = "https://httpbingo.org/json"
  body = "{}"
  req = Faraday.new(url) do |conn|
    conn.response :json, content_type: %r{application/json}
  end
  req.get('/') do |conn|
    conn.body = body
  end
end

def thread_run
  threads = []
  10.times do
    threads << Thread.new do
      run1.body.class
    end
  end
  threads.map(&:value)
end

pp thread_run
```

**Comment by ksilex** [comment]
> olleolleolle

I'm sorry, it seems like problem can occure only when requesting ClickHouse database through http method, sometimes it parsed clickhouse's json response, sometimes returned it raw (as json string)

**Comment by iMacTia** [comment]
@ksilex sorry you're having this issue, could you please add the logger middleware to your stack and check if the failing responses have anything different from the others?

```ruby
req = Faraday.new(config[:url]) do |conn|
  conn.response :logger, nil, { headers: true, bodies: true }
  conn.response :json, content_type: %r{application/json}
end
```

If you need, more info on the logger middleware are available here: https://lostisland.github.io/faraday/middleware/logger
It's also worth pointing out that ClickHouse provide a [database driver for Ruby](https://github.com/shlima/click_house), so I'm not sure if it makes sense to implement your own connector in this case. I'd suggest using the driver instead (there's even an [ActiveRecord adapter](https://github.com/PNixx/clickhouse-activerecord))

**Comment by weshatheleopard** [comment]
Having a very similar issue. Everything worked fine but after upgrading to Ruby 3.3.7 tests started failing, the culprit tacked down to `body` being returned as `String` where it previously used to be a `Hash`.

**Comment by iMacTia** [comment]
@weshatheleopard we'll need more details to be able to help, could you please open a separate issue with your Faraday connection setup and follow the instructions in [my comment above](https://github.com/lostisland/faraday/issues/1388#issuecomment-1023047393) to share the logger output?

I'm gonna close this issue as stale as it has been open for 3 years now 😄
